contracts/Strategy.sol

```diff
@@ -72,24 +143,79 @@ contract Strategy is BaseStrategy {
         // NOTE: Maintain invariant `want.balanceOf(this) >= _liquidatedAmount`
         // NOTE: Maintain invariant `_liquidatedAmount + _loss <= _amountNeeded`
 
-        uint256 totalAssets = want.balanceOf(address(this));
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+        uint256 _liquidWant = wantBalance();
+
+        _amountNeeded = Math.min(_amountNeeded, estimatedTotalAssets()); // Otherwise we can end up declaring a liquidation loss when _amountNeeded is more than we own
+
+        if (_liquidWant < _amountNeeded) {
+            _unstakeLPTokens(stakedLPBalance());
+
+            //withdraw from pool
+            _withdrawLiquidity(unstakedLPBalance());
+
+            //check current want balance
+            uint256 _postWithdrawWant = wantBalance();
+
+            //redeposit to pool
+            if (_postWithdrawWant > _amountNeeded) {
+                _addliquidity(_postWithdrawWant.sub(_amountNeeded));
+
+                //Stake the LP tokens
+                synStakingMC.deposit(pid, unstakedLPBalance());
+            }
+
+            _liquidWant = wantBalance();
+        }
+
+        if (_amountNeeded > _liquidWant) {
+            _liquidatedAmount = _liquidWant;
+            _loss = _amountNeeded.sub(_liquidWant);
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
     function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
-        return want.balanceOf(address(this));
+        synStakingMC.emergencyWithdraw(pid, address(this));
+
+        uint256 _lpTokenBalance = unstakedLPBalance();
+        if (_lpTokenBalance > 0) {
+            // Try to withdraw everything in `want`
+            _withdrawLiquidity(_lpTokenBalance);
+
+            _lpTokenBalance = unstakedLPBalance();
+
+            // If we still have LP tokens after trying to withdraw everything in `want`,
+            // withdraw in anything else that we can get
+            if (_lpTokenBalance > 0) {
+                uint256[] memory minAmounts = new uint256[](3);
+                //todo: should we try to not withdraw any amount in usdc?
+                minAmounts[0] = 0; // nUSD
```

Comment:
remove this line.
0 is default vaule. so no need to set it. in fact, it actually consumes extra gas to re-assign value to 0